src/dbnode/persist/fs/msgpack/stream_with_digest.go

```diff
@@ -0,0 +1,114 @@
+package msgpack
+
+import (
+	"errors"
+	"hash"
+	"hash/adler32"
+)
+
+var (
+	// errChecksumMismatch returned when the calculated checksum doesn't match the stored checksum
+	errChecksumMismatch = errors.New("calculated checksum doesn't match stored checksum")
+)
+
+// DecoderStreamWithDigest calculates the digest as it processes a decoder stream.
+type DecoderStreamWithDigest interface {
+	DecoderStream
+
+	// Reset resets the reader for use with a new reader.
+	Reset(stream DecoderStream)
+
+	// Digest returns the digest
+	Digest() hash.Hash32
+
+	// Validate compares the current digest against the expected digest and returns
+	// an error if they don't match.
+	Validate(expectedDigest uint32) error
+
+	// Capture provides a mechanism for manually capturing bytes to add to digest when reader is manipulated
+	// through atypical means (e.g. reading directly from the backing byte slice of a ByteReader)
+	Capture(bytes []byte) error
+
+	// Returns the decoder stream wrapped by this object
+	wrappedStream() DecoderStream
+}
+
+type decoderStreamWithDigest struct {
+	reader     DecoderStream
+	digest     hash.Hash32
+	unreadByte bool
+}
+
+func newDecoderStreamWithDigest(reader DecoderStream) DecoderStreamWithDigest {
+	return &decoderStreamWithDigest{
+		reader: reader,
+		digest: adler32.New(),
+	}
+}
+
+func (d *decoderStreamWithDigest) Read(p []byte) (n int, err error) {
+	n, err = d.reader.Read(p)
+	if n > 0 {
```

Comment:
It's kind of go-ish to deal with any short circuits first, perhaps we do that here?
```go
b, err := d.reader.Read()
if err != nil {
  return 0, err
}
if !(b > 0) {
  return b, nil
}
...
return b, nil
```